core/client/src/main/java/alluxio/client/block/stream/PacketOutStream.java

```diff
@@ -0,0 +1,281 @@
+    for (BlockWorkerClient client : clients) {
+      if (client.getWorkerNetAddress().getHost().equals(localHost)) {
+        packetWriters.add(LocalFilePacketWriter.create(client, id));
+      } else {
+        packetWriters.add(new NettyPacketWriter(context, client.getDataServerAddress(), id, length,
+            client.getSessionId(), type));
+      }
+    }
+    return new PacketOutStream(packetWriters, length);
+  }
+
+  /**
+   * Constructs a new {@link PacketOutStream} with only one {@link PacketWriter}.
+   *
+   * @param packetWriter the packet writer
+   * @param length the length of the stream
+   */
+  private PacketOutStream(PacketWriter packetWriter, long length) {
+    mLength = length;
+    mPacketWriters = new ArrayList<>(1);
+    mPacketWriters.add(packetWriter);
+    mClosed = false;
+  }
+
+  /**
+   * Constructs a new {@link PacketOutStream} with multiple {@link PacketWriter}s.
+   *
+   * @param packetWriters the packet writers
+   * @param length the length of the stream
+   */
+  private PacketOutStream(List<PacketWriter> packetWriters, long length) {
+    mLength = length;
+    mPacketWriters = packetWriters;
+    mClosed = false;
+  }
+
+  /**
+   * @return the remaining size of the block
+   */
+  @Override
+  public long remaining() {
+    long pos = Long.MAX_VALUE;
+    for (PacketWriter packetWriter : mPacketWriters) {
+      pos = Math.min(pos, packetWriter.pos());
+    }
+    return mLength - pos - (mCurrentPacket != null ? mCurrentPacket.readableBytes() : 0);
+  }
+
+  @Override
+  public void write(int b) throws IOException {
+    Preconditions.checkState(remaining() > 0, PreconditionMessage.ERR_END_OF_BLOCK);
+    updateCurrentPacket(false);
+    mCurrentPacket.writeByte(b);
+  }
+
+  @Override
+  public void write(byte[] b) throws IOException {
+    write(b, 0, b.length);
+  }
+
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    if (len == 0) {
+      return;
+    }
+
+    while (len > 0) {
+      updateCurrentPacket(false);
+      int toWrite = Math.min(len, mCurrentPacket.writableBytes());
+      mCurrentPacket.writeBytes(b, off, toWrite);
+      off += toWrite;
+      len -= toWrite;
+    }
+    updateCurrentPacket(false);
+  }
+
+  @Override
+  public void flush() throws IOException {
+    if (mClosed) {
+      return;
+    }
+    updateCurrentPacket(true);
+    for (PacketWriter packetWriter : mPacketWriters) {
+      packetWriter.flush();
+    }
+
+    // Release the channel used in the packet writer early. This is required to avoid holding the
+    // netty channel unnecessarily because the block out streams are closed after all the blocks
+    // are written.
+    if (remaining() == 0) {
+      close();
+    }
+  }
+
+  @Override
+  public void cancel() throws IOException {
+    if (mClosed) {
+      return;
+    }
+    releaseCurrentPacket();
+    for (PacketWriter packetWriter : mPacketWriters) {
+      packetWriter.cancel();
+    }
```

Comment:
For PacketOutStream, cancel is not the same as close. This is true for NettyPacketWrtier as well. I have added a comment